### ISSUE TITLE
单张图片推理车道线不匹配问题

### ISSUE THREAD

**Issue by Gannis246** [cfzd/Ultra-Fast-Lane-Detection-v2#28]
@cfzd 
if __name__ == "__main__":

    torch.backends.cudnn.benchmark = True

    args, cfg = merge_config()
    cfg.batch_size = 1
    print('setting batch_size to 1 for demo generation')

    dist_print('start testing...')
    assert cfg.backbone in ['18', '34', '50', '101', '152', '50next', '101next', '50wide', '101wide']

    if cfg.dataset == 'CULane':
        cls_num_per_lane = 18
    elif cfg.dataset == 'Tusimple':
        cls_num_per_lane = 56
    else:
        raise NotImplementedError

    net = get_model(cfg)

    state_dict = torch.load(cfg.test_model, map_location='cpu')['model']
    compatible_state_dict = {}
    for k, v in state_dict.items():
        if 'module.' in k:
            compatible_state_dict[k[7:]] = v
        else:
            compatible_state_dict[k] = v

    net.load_state_dict(compatible_state_dict, strict=False)
    net.eval()

    img_transforms = transforms.Compose([
        transforms.ToPILImage(),
        transforms.Resize((cfg.train_height, cfg.train_width)),
        transforms.ToTensor(),
        transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
    ])

    img_path = "D:/autodrive/images/00010.png"
    img = cv2.imread(img_path)
    img_h, img_w = img.shape[0], img.shape[1]
    im0 = img.copy()
    img = img_transforms(img)
    img = img.to('cuda:0')
    img = torch.unsqueeze(img, 0)
    with torch.no_grad():
        pred = net(img)
    coords = pred2coords(pred, cfg.row_anchor, cfg.col_anchor, original_image_width=img_w,
                            original_image_height=img_h)
    for lane in coords:
        for coord in lane:
            cv2.circle(im0, coord, 5, (0, 255, 0), -1)
    cv2.imshow('demo', im0)
    cv2.waitKey(0)”
这是我改的单张图片的推理，但是检测出的点与车道线匹配不上，似乎是你在加载图片是进行过仿射变换，可以解答一下吗

**Comment by cfzd** [comment]
@Gannis246 
我在另外一个issue已经回复了，你可以参考那个回复。

**Comment by Gannis246** [comment]
太感谢了，我看了很久，就是没明白cfg.crop_size这个参数是什么意思，这下懂了



------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2022年9月1日(星期四) 晚上8:00
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [cfzd/Ultra-Fast-Lane-Detection-v2] 单张图片推理车道线不匹配问题 (Issue #28)





 
@Gannis246
 我在另外一个issue已经回复了，你可以参考那个回复。

**Comment by cfzd** [comment]
https://github.com/cfzd/Ultra-Fast-Lane-Detection-v2/issues/18#issuecomment-1234179888

**Comment by Gannis246** [comment]
@Aruen24 
正常了
图片预处理部分改成这样
```
    img_transforms = transforms.Compose([
        transforms.ToPILImage(),
        transforms.Resize((int(cfg.train_height/cfg.crop_ratio), cfg.train_width)),
        transforms.ToTensor(),
        transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
    ])
    im0 = img.copy()
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    img_h, img_w = img.shape[0], img.shape[1]
    img = img_transforms(img)
    img = img[:, -cfg.train_height:, :]
    img = img.to('cuda:0')
    img = torch.unsqueeze(img, 0)

**Comment by Aruen24** [comment]
![frame_0](https://user-images.githubusercontent.com/27750891/188045743-035598e0-5807-437d-bb04-5a4a2a95eeec.jpg)
测试一下这张图片看看是否正常

**Comment by Gannis246** [comment]
@Aruen24 
效果还是可以的
![11](https://user-images.githubusercontent.com/69740611/188047319-b7e42f66-b8fc-40bb-82f3-4203464622cd.jpg)

**Comment by Aruen24** [comment]
@Gannis246 你是自己训的模型还是？我自己训练的测试的效果不好

**Comment by Gannis246** [comment]
@Aruen24 我就是用的预训练模型。感觉这个模型泛化性不是很好，近处车道遮挡，压线行驶，大弯道检测不出来，挡风玻璃反光也会偏移

**Comment by Aruen24** [comment]
@Gannis246 
`import torch, os, cv2
from utils.dist_utils import dist_print
import torch, os
from utils.common import merge_config, get_model
import tqdm
import torchvision.transforms as transforms
from data.dataset import LaneTestDataset

def pred2coords(pred, row_anchor, col_anchor, local_width = 1, original_image_width = 1640, original_image_height = 590):
    batch_size, num_grid_row, num_cls_row, num_lane_row = pred['loc_row'].shape
    batch_size, num_grid_col, num_cls_col, num_lane_col = pred['loc_col'].shape

    max_indices_row = pred['loc_row'].argmax(1).cpu()
    # n , num_cls, num_lanes
    valid_row = pred['exist_row'].argmax(1).cpu()
    # n, num_cls, num_lanes

    max_indices_col = pred['loc_col'].argmax(1).cpu()
    # n , num_cls, num_lanes
    valid_col = pred['exist_col'].argmax(1).cpu()
    # n, num_cls, num_lanes

    pred['loc_row'] = pred['loc_row'].cpu()
    pred['loc_col'] = pred['loc_col'].cpu()

    coords = []

    row_lane_idx = [1,2]
    col_lane_idx = [0,3]

    for i in row_lane_idx:
        tmp = []
        if valid_row[0,:,i].sum() > num_cls_row / 2:
            for k in range(valid_row.shape[1]):
                if valid_row[0,k,i]:
                    all_ind = torch.tensor(list(range(max(0,max_indices_row[0,k,i] - local_width), min(num_grid_row-1, max_indices_row[0,k,i] + local_width) + 1)))
                    
                    out_tmp = (pred['loc_row'][0,all_ind,k,i].softmax(0) * all_ind.float()).sum() + 0.5
                    out_tmp = out_tmp / (num_grid_row-1) * original_image_width
                    tmp.append((int(out_tmp), int(row_anchor[k] * original_image_height)))
            coords.append(tmp)

    for i in col_lane_idx:
        tmp = []
        if valid_col[0,:,i].sum() > num_cls_col / 4:
            for k in range(valid_col.shape[1]):
                if valid_col[0,k,i]:
                    all_ind = torch.tensor(list(range(max(0,max_indices_col[0,k,i] - local_width), min(num_grid_col-1, max_indices_col[0,k,i] + local_width) + 1)))
                    
                    out_tmp = (pred['loc_col'][0,all_ind,k,i].softmax(0) * all_ind.float()).sum() + 0.5

                    out_tmp = out_tmp / (num_grid_col-1) * original_image_height
                    tmp.append((int(col_anchor[k] * original_image_width), int(out_tmp)))
            coords.append(tmp)

    return coords
if __name__ == "__main__":
    torch.backends.cudnn.benchmark = True

    args, cfg = merge_config()
    cfg.batch_size = 1
    print('setting batch_size to 1 for demo generation')

    dist_print('start testing...')
    assert cfg.backbone in ['18', '34', '50', '101', '152', '50next', '101next', '50wide', '101wide']

    if cfg.dataset == 'CULane':
        cls_num_per_lane = 18
    elif cfg.dataset == 'Tusimple':
        cls_num_per_lane = 56
    else:
        raise NotImplementedError

    net = get_model(cfg)

    state_dict = torch.load(cfg.test_model, map_location='cpu')['model']
    compatible_state_dict = {}
    for k, v in state_dict.items():
        if 'module.' in k:
            compatible_state_dict[k[7:]] = v
        else:
            compatible_state_dict[k] = v

    net.load_state_dict(compatible_state_dict, strict=False)
    net.eval()

    img_transforms = transforms.Compose([
        transforms.ToPILImage(),
        transforms.Resize((int(cfg.train_height/cfg.crop_ratio), cfg.train_width)),
        transforms.ToTensor(),
        transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
    ])

    img_path = "./frame_0.jpg"
    img = cv2.imread(img_path)
    im0 = img.copy()
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    img_h, img_w = img.shape[0], img.shape[1]
    img = img_transforms(img)
    img = img[:, -cfg.train_height:, :]
    img = img.to('cuda:0')
    img = torch.unsqueeze(img, 0)
    with torch.no_grad():
        pred = net(img)
    coords = pred2coords(pred, cfg.row_anchor, cfg.col_anchor, original_image_width=img_w,
                            original_image_height=img_h)
    for lane in coords:
        for coord in lane:
            cv2.circle(im0, coord, 5, (0, 255, 0), -1)
    #cv2.imshow('demo', im0)
    cv2.imwrite('out.jpg', im0)
    #cv2.waitKey(0)`
![out](https://user-images.githubusercontent.com/27750891/188068670-4e972a3a-828f-48c1-adf8-7f003e65ab46.jpg)
I detect yhe result. the code has any problem?

**Comment by Gannis246** [comment]
@Aruen24 
the code has no problem. if this model is trained by yourself, maybe you have changed the `cfg.train_height` or `cfg.train_height`.

**Comment by Aruen24** [comment]
@Gannis246 
I use the pretrained model tusimple_res34.pth. But get the same result.

**Comment by Gannis246** [comment]
@Aruen24 
I have run every pretrained model and I think culane_res18.pth and culane_res34.pth are best

**Comment by Aruen24** [comment]
@Gannis246 
so you use my code with tusimple_res34.pth pretrained model test the picture.  what is the test result?

**Comment by Gannis246** [comment]
@Aruen24 
![11](https://user-images.githubusercontent.com/69740611/188101582-42e54c46-3c20-4280-82db-f56dc532103a.jpg)
like this

**Comment by Gannis246** [comment]
@Aruen24 
代码没问题的，其他文件我都没改你就用culane_res18试一下，不行的话换张图试试

**Comment by Yutong-gannis** [comment]
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "cfzd/Ultra-Fast-Lane-Detection-v2"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年9月2日(星期五) 上午10:29
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [cfzd/Ultra-Fast-Lane-Detection-v2] 单张图片推理车道线不匹配问题 (Issue #28)





 

 测试一下这张图片看看是否正常

**Comment by guodaoyi** [comment]
@Yutong-gannis 请问基于tusimple的训练模型车道线检测飘到天上了，这个问题您解决了吗

**Comment by zhongwei1122** [comment]
> @Aruen24 效果还是可以的 ![11](https://user-images.githubusercontent.com/69740611/188047319-b7e42f66-b8fc-40bb-82f3-4203464622cd.jpg)

请问为什么按照你的代码修改demo.py，基于预训练模型测试tusimple中的单张图片，检测出的点和车道线匹配不上
